Add answers for initial FAQ (#943)
graphql/graphql.github.io: src/content/faq/General.md

## What is a GraphQL client and why would I use one?

GraphQL clients can help you handle [queries, mutations,](/learn/queries/) and [subscriptions](https://spec.graphql.org/draft/#sec-Subscription) to a [GraphQL server](https://www.howtographql.com/advanced/1-server/). They use the underlying structure of a GraphQL API to automate certain processes. This includes batching, UI updates, build-time schema validation, and more.

A [list of GraphQL clients](/code/#graphql-clients) in various languages is available on our Code page. There’s also an [in-depth explanation of their benefits](https://www.howtographql.com/advanced/0-clients/) on How To GraphQL.

You don't need a specific client to work with GraphQL, though. You might want to start out by [issuing GraphQL results with a regular HTTP client](/learn/serving-over-http/). Then later switch to a GraphQL-optimized client as your application grows in complexity.

## Is GraphQL owned by Facebook?

No, GraphQL is governed by the [GraphQL Foundation](#what-is-the-graphql-foundation).

> That said, the specification was originally developed at Facebook and [Facebook is a member](https://foundation.graphql.org/members/) of the GraphQL Foundation. You might notice that some of our [GitHub repositories](https://github.com/graphql/) still have the license listed under Facebook Inc. We're updating those and have already converted major projects, like [GraphiQL](https://github.com/graphql/graphiql/blob/main/LICENSE) and [DataLoader](https://github.com/graphql/dataloader/blob/master/LICENSE), to the MIT License.

Reviewer: It's not the license but the licensor that this question relates to; we've been changing `(c) Facebook` to `(c) GraphQL Contributor` so that's what we should note:

```suggestion
That said, the specification was originally developed at Facebook and [Facebook is a member](https://foundation.graphql.org/members/) of the GraphQL Foundation. You might notice that some of our [GitHub repositories](https://github.com/graphql/) still have the license listed under Facebook Inc. We're updating those and have already converted major projects, like [GraphiQL](https://github.com/graphql/graphiql/blob/main/LICENSE) and [DataLoader](https://github.com/graphql/dataloader/blob/master/LICENSE), to the the new copyright: "Copyright (c) 2020 GraphQL Contributors."
```

@brianwarner Please can you confirm if https://github.com/graphql/dataloader/blob/249b2b966a8807c50e07746ff04acb8c48fa4357/LICENSE#L4 is out of date; it should be "GraphQL Contributors" rather than "GraphQL Foundation," right?

Author: Good call, I'll update!

For more practical guides, visit the [How to GraphQL](https://www.howtographql.com/) fullstack tutorial website. We also have a free online course with edX, [Exploring GraphQL: A Query Language for APIs](https://www.edx.org/course/exploring-graphql-a-query-language-for-apis). 

Before you start your learning journey, make sure you know what an API is and how communication generally works between client and server.

## Is GraphQL frontend or backend?

Both. GraphQL specifies how you can [exchange information between client and server](https://www.howtographql.com/basics/3-big-picture/). This includes how the server can indicate [what data and operations are available](/learn/introspection/), how the client should [format requests](/learn/queries/), how the server should [execute these queries](/learn/execution/), and what the client will [receive in response](/learn/serving-over-http/#response).

## Is GraphQL only for React or JavaScript developers?

> No, not at all. [GraphQL services can be written in any language](/learn/schema/#type-language). Our [Code page](/code/) contains a long list of libraries in many different programming languages to help with that.

Reviewer 2: > No, not at all.

I would add that GraphQL is the specification that can be implemented in any language

Author: Do you mean in replacement of `GraphQL services can be written in any language` or would you want to somehow combine the two?

You don't need a specific client to work with GraphQL, though. You might want to start out by [issuing GraphQL results with a regular HTTP client](/learn/serving-over-http/). Then later switch to a GraphQL-optimized client as your application grows in complexity.

## Is GraphQL owned by Facebook?

No, GraphQL is governed by the [GraphQL Foundation](#what-is-the-graphql-foundation).

That said, the specification was originally developed at Facebook and [Facebook is a member](https://foundation.graphql.org/members/) of the GraphQL Foundation. You might notice that some of our [GitHub repositories](https://github.com/graphql/) still have the license listed under Facebook Inc. We're updating those and have already converted major projects, like [GraphiQL](https://github.com/graphql/graphiql/blob/main/LICENSE) and [DataLoader](https://github.com/graphql/dataloader/blob/master/LICENSE), to the the new copyright: "Copyright (c) 2020 GraphQL Contributors."

## What is the GraphQL Foundation?

> The [GraphQL Foundation](https://foundation.graphql.org/faq/) is a neutral foundation that provides governance for GraphQL. This includes vendor-neutral oversight of open-source repositories, funding, events, and more. It's hosted under the [Linux Foundation](https://www.linuxfoundation.org/) and consists of [representatives from 20 different companies](https://foundation.graphql.org/members/). The idea is that it’s an impartial and open home for the GraphQL community.

Reviewer 2: I think we should avoid exact numbers since it will be outdated frequently.
How about "dozens of different companies" or something similar?

Author: Good call, I'll update!

---

## Why should I use GraphQL?

It depends on your use case, but in general, GraphQL has a few key features that stand out. For example, GraphQL enables you to:

* Improve performance by [sending queries](/learn/queries/) for the exact data you need. 
* Aggregate data from [multiple UI components](/learn/queries/#fragments) in [a single request](/learn/best-practices/#server-side-batching-caching ).
* Create a representation of your data that feels familiar and natural ([a graph](/learn/thinking-in-graphs/#it-s-graphs-all-the-way-down)).
* Ensure that all of your data is [statically typed](/learn/schema/) and these types inform [what queries the schema supports](/learn/introspection/).
* [Reduce the need for breaking changes](/learn/best-practices/#versioning), but utilize a [built-in mechanism for deprecations](https://spec.graphql.org/draft/#sec-Deprecation) when you need to.

> * Examine queries and responses through community tools like [GraphiQL](https://github.com/graphql/graphiql).

Author: Something else I noticed: This is a lot of bullet points 🤪  Since we mention that there are more reasons outlined on the homepage, I'm debating cutting point 1, 2, 4 and/or 6.

 I'm thinking 6 (community tools) and 3 (graph-like data). Wdyt @IvanGoncharov / @benjie?

Reviewer: I've tried to come up with a list of reasons I like GraphQL without looking at your list above, and then place them into the right order for me. It's very hard to order them!

- Every GraphQL API follows same spec: tooling (e.g. GraphiQL, codegen, etc) is very powerful and shared - worth investing in; learn once, apply everywhere
- GraphQL client synchronises data across my interface with virtually no effort
- Data's naturally in the right shape for the client, no/minimal post-processing required
- Strong typing
- Autogenerated client code (graphql-code-gen, relay compiler, etc)
- Data requirements co-located with data consumers (i.e. fragments)
- No API versioning (@deprecated, etc)
- Editor integration
- Linting integration

Reviewer: I think the shared precise specification is probably the number 1 biggest reason to use GraphQL because so much of the rest of GraphQL's benefits flow from this one thing. You don't have to build your own tooling, everyone can use the same tooling!

I agree with dropping 1, I'd drop the batching from 2, I think 4 should be kept, re: point 6 tooling is incredible, but maybe it could be rewritten to something like "Incredibly rich and powerful tooling ecosystem, with GUIs (GraphiQL, Altair, GraphQL Playground, ...), Editor Integration (VSCode, Vim, ...) code generation (gql-codegen, Relay, Apollo, ...), linting, analytics (Apollo Optics), and much more". I suspect removing the parenthesis and just linking the entire point to https://graphql.org/code/ would make sense.

For an opinionated perspective on how GraphQL compares to REST, check out [How To GraphQL](https://www.howtographql.com/basics/1-graphql-is-the-better-rest/).

## How can I learn GraphQL?

There are many resources available to help you learn GraphQL, including this website. In [our documentation](/learn/), you’ll find a series of articles that explain essential GraphQL concepts and how they work. Our [Community page](/community) is full of resources to reference and groups to join.

For more practical guides, visit the [How to GraphQL](https://www.howtographql.com/) fullstack tutorial website. We also have a free online course with edX, [Exploring GraphQL: A Query Language for APIs](https://www.edx.org/course/exploring-graphql-a-query-language-for-apis). 

## Is GraphQL frontend or backend?

Neither. Instead, it’s [how you can exchange information between the two](https://www.howtographql.com/basics/3-big-picture/). Both the client and the server should understand and communicate through the GraphQL specification. 

> That said, GraphQL describes a specific way to [execute a query](/learn/execution/), which typically happens on the backend.

Reviewer: I'm a bit uncomfortable with this; I feel like the execution (on the backend) is a significant component of GraphQL. I wonder if it could be more of a "both" rather than "neither"? Maybe something like:

> GraphQL specifies how you can exchange information between client and server, including how the server can indicate what data and operations are available [link to introspection], how the client should format requests [link to query language], how the server should execute these queries [link to execution], and what the client will receive in response [link to "response is same shape as query"].

Author: Absolutely - I really like that, so I'll update!

The only thing I'm hung up on is that I think FAQs are most effective when they answer the question right away... could it make sense to literally say like `It interacts with both, actually.` (or something better lol) at the beginning?

Reviewer: I think just starting with `Both.` instead of `Neither.` might be sufficient.

## What is a GraphQL client and why would I use one?

GraphQL clients can help you handle [queries, mutations,](/learn/queries/) and [subscriptions](https://spec.graphql.org/draft/#sec-Subscription) to a [GraphQL server](https://www.howtographql.com/advanced/1-server/). They use the underlying structure of a GraphQL API to automate certain processes. This includes batching, UI updates, build-time schema validation, and more.

A [list of GraphQL clients](/code/#graphql-clients) in various languages is available on our Code page. There’s also an [in-depth explanation of their benefits](https://www.howtographql.com/advanced/0-clients/) on How To GraphQL.

You don't need a specific client to work with GraphQL, though. You might want to start out by [issuing GraphQL results with a regular HTTP client](/learn/serving-over-http/). Then later switch to a GraphQL-optimized client as your application grows in complexity.

## Is GraphQL owned by Facebook?

No, GraphQL is governed by the [GraphQL Foundation](#what-is-the-graphql-foundation).

> That said, the specification was originally developed at Facebook and [Facebook is a member](https://foundation.graphql.org/members/) of the GraphQL Foundation. Some of our [repositories on GitHub](https://github.com/graphql/) still have the license listed under Facebook Inc, but we're updating those.

Reviewer: Is it worth noting that the higher profile projects have already been converted?

Author: Which projects are you thinking of?

Reviewer: I was thinking https://github.com/graphql/graphql-js/blob/master/LICENSE, https://github.com/graphql/express-graphql/blob/master/LICENSE, https://github.com/graphql/graphiql/blob/main/LICENSE, https://github.com/graphql/dataloader/blob/master/LICENSE; I was also thinking the Spec, but it doesn't seem to be so clear there, and the OWFa1.0 is signed on behalf of facebook (<https://github.com/graphql/graphql-spec/blob/master/signed-agreements/Facebook-GraphQL-OWFa1.0.pdf>) so... :shrug:

## Why should I use GraphQL?

It depends on your use case, but in general, GraphQL has a few key features that stand out. For example, GraphQL enables you to:

> * Improve performance and stability by [sending queries](/learn/queries/) for the exact data you need. 

Reviewer 3: I am not sure what stability means here. Can we either explain it better or remove it?

Author: Removed!

* Ensure that all of your data is [statically typed](/learn/schema/) and these types inform [what queries the schema supports](/learn/introspection/).
* [Reduce the need for breaking changes](/learn/best-practices/#versioning), but utilize a [built-in mechanism for deprecations](https://spec.graphql.org/draft/#sec-Deprecation) when you need to.
* Examine queries and responses through community tools like [GraphiQL](https://github.com/graphql/graphiql).

And many others! [Our homepage](/) outlines some more convincing reasons to use GraphQL.

Additionally, it doesn’t take rewriting your entire application to try out GraphQL. You can even start with a single HTTP request that wraps an existing REST call, for instance. Your [GraphQL schema](/learn/thinking-in-graphs/#shared-language) and [business domain model](/learn/thinking-in-graphs/#business-logic-layer) can be expanded gradually, so we’d recommend focusing on one use case at first and only build the part of the schema you need for that to work. 

## Is GraphQL a database language like SQL?

No, but this is a common misconception.

> GraphQL is a specification for how to request and modify data. That data can be provided by [arbitrary functions](/learn/) and those functions can call APIs, databases, [local cache](/learn/caching/) or any representation of data. The main and most common use case is calling APIs.

Reviewer 3: Databases are almost always exposed via some form of API, so I am not sure this differentiation makes sense. 
Based on that I am also not sure I agree with `main and most common use case is calling APIs`: Getting data from a database is very common, maybe the most common use case in my experience.

Author: I see what you mean. The API use case is the one that is most highlighted on the homepage and promoted in other internal documents, so that's why I felt like it was important to distinguish here.

I'd be curious to hear how you'd distinguish GraphQL from SQL - mostly because then I can incorporate that into this answer 😂

Reviewer 3: I am not sure I have a great answer to SQL vs GraphQL, but I think there are two fundamental aspects if you really want to compare both:

1. Where the data comes from: GraphQL is agnostic to the actual datasource: GraphQL is a data aggregator across (possibly) multiple arbitrary and therefore heterogeneous data sources. SQL is a query language for a single relational database.  

2. "Distance" between client and server or visibility of the API: GraphQL is a query language between a client and server where the client is normally  "far" away. The client is normally a web/mobile/native client or the client can be even arbitrary for public GraphQL APIs. Compare that with SQL: normally SQL is an implementation detail of a backend server and not in any way exposed to a frontend or to the public.  

These are general statements describing the most common and best practices: there are exceptions to nearly everything said above (e.g. some are using SQL statements directly in the frontend, some are using GraphQL to access their database instead of SQL etc).

I hope that helps. :-)

Reviewer: I'm not sure what I've written here is the best FAQ answer (also it's a bit long), but maybe it'll help inspire something better?

---

No, but this is a common misconception.

GraphQL is a specification typically used for remote client-server communications, and is agnostic to the data source(s) used to retrieve data/persist changes. Data access/manipulation is performed via [arbitrary functions](/learn/) called "resolvers"; GraphQL coordinates and aggregates the data from these resolver functions, returning the result to the client. General advice is that these resolver functions should delegate to a "business logic layer", written in your organizations' preferred programming language, responsible for communicating with the various underlying data sources (remote APIs, databases, [local cache](/learn/caching/), and nearly anything else your programming language can access).

Author: Both of these are _super_ helpful, thank you @andimarek and @benjie! I'm going to take both and hopefully come up with an answer that covers everything (but isn't too long).

Author: Alright I tried, curious what you think 😁

GraphQL and REST can actually co-exist in your stack. For example, you can abstract REST APIs behind a [GraphQL server](https://www.howtographql.com/advanced/1-server/). This can be done by masking your REST endpoint into a GraphQL endpoint using [root resolvers](/learn/execution/#root-fields-resolvers). 

For an opinionated perspective on how GraphQL compares to REST, check out [How To GraphQL](https://www.howtographql.com/basics/1-graphql-is-the-better-rest/).

## How can I learn GraphQL?

There are many resources available to help you learn GraphQL, including this website. In [our documentation](/learn/), you’ll find a series of articles that explain essential GraphQL concepts and how they work. Our [Community page](/community) is full of resources to reference and groups to join.

For more practical guides, you can visit the [How to GraphQL](https://www.howtographql.com/) fullstack tutorial website. We also have a free online course with edX, [Exploring GraphQL: A Query Language for APIs](https://www.edx.org/course/exploring-graphql-a-query-language-for-apis). 

## Is GraphQL frontend or backend?

> Neither. Instead, it’s [how you can exchange information between the two](https://www.howtographql.com/basics/3-big-picture/). Both the client and the server should understand and communicate through the GraphQL specification.

Reviewer 3: While I agree with this statement, I think it is fair to call out that GraphQL describes a specific way to execute a query, which normally happens in the backend.

Author: Do you think it would be enough to add a sentence about that within this answer? Or do you think it'd be more productive to reword the entire answer?

Reviewer 3: Adding it would address my concerns. No need to reword it completely.

## Does GraphQL replace Redux or other state management libraries?

> No, GraphQL isn’t a state management library - but it can reduce the need for one.

Reviewer 3: I am not sure we should add this very specific JS question to the FAQ. We are not helping here to improve the situation described in the next question.

Author: Do you think removing `Redux` would help? This is a _super_ common question, so I believe it should be included in the FAQ eventually - but we could hold it until there are more subsections.

Reviewer 3: I think this is a bit a tough one: I totally get that this is a super common question ... if there would be multiple sections I would recommend to move it.

Maybe we can find a pragmatic solution in the mean time: prefix it with "JS" or so? I just want to reflect a bit more clear that GraphQL is a big ecosystem which has a lot of things to offer outside JS. I think especially for beginners this is important to see and at least get a nudge like "there is more to it than JS". Hope that makes sense.

Author: Absolutely makes sense! One of the primary goals for this project is to `Build a resource for everyone, regardless of programming language` so I definitely want to make sure that it aligns. 

I'll remove this question from this PR and keep it in the backlog for once we create subsections within the FAQ 😊

## Is GraphQL only for React or JavaScript developers?

No, not at all. [GraphQL services can be written in any language](/learn/schema/#type-language). Our [Code page](/code/) contains a long list of libraries in many different programming languages to help with that.

It’s understandable why you’d think this, though. GraphQL was introduced at a [React conference](https://www.youtube.com/watch?v=9sc8Pyc51uU) and [GraphQL.js](/graphql-js/) is one of the most widely used reference implementations to date. We know this can be confusing, so we’re working to improve our documentation and add more code samples that aren’t written in JavaScript.

## What is a GraphQL client and why would I use one?

> GraphQL clients can help you handle [queries, mutations,](/learn/queries/) and [subscriptions](https://spec.graphql.org/draft/#sec-Subscription) to a [GraphQL server](https://www.howtographql.com/advanced/1-server/). This could be done manually, but it would require a lot of work depending on the amount of data you have. By utilizing the underlying structure of a GraphQL API, clients can abstract away some of these tedious processes. Features that can be automated include batching, caching, UI updates, build-time schema validation, and many more. 

Reviewer 3: I think this section is misleading: one of the great strength of GraphQL is actually the opposite: you don't need a specific client. It is perfectly fine to use existing HTTP clients. While it is true you can do more it is in no way required.

Author: Totally agree. Do you think it'd be helpful to distinguish this earlier on in the answer? ie. mention right away that they aren't _necessary_ then explain a few of the reasons why people choose to work with them?

Author: Honestly might have made this answer worse 😂  Any suggestions appreciated @andimarek @benjie 🙏🏼

You can find a list of GraphQL clients in various languages on our [Code page](/code/#graphql-clients). There’s also an [in-depth explanation of their benefits](https://www.howtographql.com/advanced/0-clients/) on How To GraphQL.

> It takes more time to implement a GraphQL client upfront, but it becomes worth it the more features you need. You might want to start out [using HTTP requests](/learn/serving-over-http/) as the underlying transport layer (here’s [an example in JavaScript](/graphql-js/graphql-clients/)), then switch to a more intricate client service as your application grows in complexity.

Reviewer 3: I am not sure what "more time to implement a GraphQL client upfront" means.

Author: I meant like... GraphQL clients come with many features out-of-the-box and it can be difficult to figure out what you need when you're first working with them. I'd argue that implementing a GraphQL client (Apollo for instance) is more complicated than doing HTTP requests initially. But I can see why it'd be confusing in this context. 

Do you think it'd make sense to cut this sentence and start with `You might want to start...`? This could address both this comment and the previous.

Reviewer: I think cutting the sentence makes sense.

Separately, "using HTTP requests as the underlying transport layer" doesn't feel clear enough to me (since most of the GraphQL clients also use HTTP requests as the underlying transport layer), can we somehow make this more obvious that we mean issuing HTTP requests manually? Maybe something like: "You might want to start out by issuing GraphQL results with a regular HTTP client, later switching to a GraphQL optimised client as your application grows in complexity."

(Sorry for my UK English :wink:)

Author: So much better - thank you @benjie!

## What is the GraphQL Foundation?

> The [GraphQL Foundation](https://foundation.graphql.org/faq/) provides governance for GraphQL as well as vendor-neutral oversight of funding, events, operations resources, and more. It was formed in 2018 by [various tech companies](https://landscape.graphql.org/category=graph-ql-foundation-member&format=logo-mode) and hosted under the [Linux Foundation](https://www.linuxfoundation.org/). The idea is that it’s an open, neutral home for the GraphQL community.

Reviewer 3: Why do we highlight the members who initially joined? Can we link to all current members here?

Author: > Why do we highlight the members who initially joined?

I think the context of who it was formed by is important, especially as it determines who has influenced the development of GraphQL overtime. However, I agree that adding a question like `Is GraphQL owned by Facebook?` would make this less relevant.

> Can we link to all current members here?

Yes, I can do that!

Author: I changed it to only include the current members (the list in the same, but hopefully the corresponding text is more clear). Curious what you think!

## Why should I use GraphQL?

It depends on your use case, but in general, GraphQL has a few key features that stand out. For example, GraphQL enables you to:

* Improve performance by [sending queries](/learn/queries/) for the exact data you need. 
* Aggregate data from [multiple UI components](/learn/queries/#fragments) in [a single request](/learn/best-practices/#server-side-batching-caching ).
* Create a representation of your data that feels familiar and natural ([a graph](/learn/thinking-in-graphs/#it-s-graphs-all-the-way-down-https-en-wikipedia-org-wiki-turtles-all-the-way-down)).
* Ensure that all of your data is [statically typed](/learn/schema/) and these types inform [what queries the schema supports](/learn/introspection/).
* [Reduce the need for breaking changes](/learn/best-practices/#versioning), but utilize a [built-in mechanism for deprecations](https://spec.graphql.org/draft/#sec-Deprecation) when you need to.
* Examine queries and responses through community tools like [GraphiQL](https://github.com/graphql/graphiql).

> And many others! [Our homepage](/) outlines some more convincing reasons to use GraphQL.

Reviewer: I feel like if the homepage has "more convincing reasons" then they should be here (at the top of the list). Instead, perhaps:

```suggestion
[Our homepage](/) outlines even more reasons to use GraphQL.
```

> Additionally, it doesn’t take rewriting your entire application to try out GraphQL. You can even start with a single HTTP request that wraps an existing REST call, for instance. Your [GraphQL schema](/learn/thinking-in-graphs/#shared-language) and [business domain model](/learn/thinking-in-graphs/#business-logic-layer) can be expanded gradually, so we’d recommend focusing on one use case at first and only build the part of the schema you need for that to work. 

Reviewer: I've condensed this a little; feel free to dismiss.

```suggestion
You can try out GraphQL without rewriting your entire application; for example: start with a single HTTP request that wraps an existing REST call. Your [GraphQL schema](/learn/thinking-in-graphs/#shared-language) and [business domain model](/learn/thinking-in-graphs/#business-logic-layer) can be expanded gradually; we recommend focusing on one use case at first and only building the part of the schema needed for that. 
```

Author: Definitely agree with condensing but confession: I hate semicolons 😂  So I'll apply the changes - just without those 😁

## Is GraphQL a database language like SQL?

No, but this is a common misconception.

GraphQL is a specification for how to request and modify data. That data can be provided by [arbitrary functions](/learn/) and those functions can call APIs, databases, [local cache](/learn/caching/) or any representation of data. The main and most common use case is calling APIs.

GraphQL is frequently used for database interactions though. For example, it can be built into a wrapper around a database or a [GraphQL server](https://www.howtographql.com/advanced/1-server/) can query various databases. There are also [services](/code/#services) that implement the GraphQL specification and then allow clients to use GraphQL to query that data source. But it can also be used without a database at all.

For more information on how to get GraphQL to interact with your database, check out our [documentation on resolvers](/learn/execution/#root-fields-resolvers).

## Does GraphQL replace REST?

> No, not necessarily. They both handle APIs and [serve the same purpose](/learn/thinking-in-graphs/#business-logic-layer) from a business perspective. GraphQL is considered an alternative to REST, but it’s not a definitive replacement.

Reviewer: I think saying that GraphQL and REST serve the same purpose is an over-generalisation; the scope of GraphQL is much broader than that of REST in my opinion, take Gatsby's usage of GraphQL for example. I've softened the meaning:

```suggestion
No, not necessarily. They both handle APIs and can [serve similar purposes](/learn/thinking-in-graphs/#business-logic-layer) from a business perspective. GraphQL is often considered an alternative to REST, but it’s not a definitive replacement.
```

For an opinionated perspective on how GraphQL compares to REST, check out [How To GraphQL](https://www.howtographql.com/basics/1-graphql-is-the-better-rest/).

## How can I learn GraphQL?

There are many resources available to help you learn GraphQL, including this website. In [our documentation](/learn/), you’ll find a series of articles that explain essential GraphQL concepts and how they work. Our [Community page](/community) is full of resources to reference and groups to join.

For more practical guides, you can visit the [How to GraphQL](https://www.howtographql.com/) fullstack tutorial website. We also have a free online course with edX, [Exploring GraphQL: A Query Language for APIs](https://www.edx.org/course/exploring-graphql-a-query-language-for-apis). 

## Is GraphQL frontend or backend?

Neither. Instead, it’s [how you can exchange information between the two](https://www.howtographql.com/basics/3-big-picture/). Both the client and the server should understand and communicate through the GraphQL specification.

## Does GraphQL replace Redux or other state management libraries?

No, GraphQL isn’t a state management library - but it can reduce the need for one.

> One benefit of state management libraries like Redux is that they can manipulate API responses into a format that your application understands. With GraphQL, you have control over [what data you request](/learn/queries/#fields) and [how that data is formatted](/learn/execution/). So this benefit is already built-in. Many [client libraries](https://graphql.org/code/#graphql-clients) can also be used to manage state and have features like caching built-in. You may still decide to implement a state management library, but using it to format response data won’t be necessary.

Reviewer: > and [how that data is formatted](/learn/execution/)
> [...]
> using it to format response data won’t be necessary

We need to be careful not to over-sell this, a lot of GraphQL beginners that I've spoken with get frustrated that they cannot request GraphQL results in custom formats (e.g. flattening trees, forming a key-value map, etc). Sure this could be achieved with things such as [GraphQL-Lodash](https://github.com/APIs-guru/graphql-lodash) but I don't think that's something we can say is intrinsic to GraphQL.

Maybe instead we can go with something softer, such as:

```suggestion
One benefit of state management libraries like Redux is that they can manipulate API responses into a format that your application understands. With GraphQL, you have control over [what data you request](/learn/queries/#fields), and typically results are formatted in a client-friendly way by virtue of the graph design, so this benefit is already built-in. Many [client libraries](https://graphql.org/code/#graphql-clients) can also be used to manage state and have features like caching built-in. You may still decide to implement a state management library, but using it to format response data is generally not necessary.
```

Author: So given the previous discussion with @andimarek, I've decided to put this question in the backlog and not include it in this PR. _However_ your notes are great and I've applied those changes to the existing answer so that they are ready whenever we do include it 🙏🏼

## Is GraphQL only for React or JavaScript developers?

No, not at all. [GraphQL services can be written in any language](/learn/schema/#type-language). Our [Code page](/code/) contains a long list of libraries in many different programming languages to help with that.

> It’s understandable why you’d think this, though. GraphQL was introduced at a [React conference](https://www.youtube.com/watch?v=9sc8Pyc51uU) and [GraphQL.js](/graphql-js/) is one of the most widely used reference implementations to date. We know this can be confusing, so we’re working to improve our documentation and add more code samples that aren’t written in JavaScript.

Reviewer: Nit: GraphQL.js is the _reference_ implementation; I think all the other implementations are simply "implementations".

```suggestion
It’s understandable why you’d think this, though. GraphQL was introduced at a [React conference](https://www.youtube.com/watch?v=9sc8Pyc51uU) and [GraphQL.js](/graphql-js/) is one of the most widely used implementations to date. We know this can be confusing, so we’re working to improve our documentation and add more code samples that aren’t written in JavaScript.
```
